File: python/progress_bar/progress_bar.py
# -*- coding: utf-8 -*-

import sys


class ProgressBar(object):
    """Output a simple progress bar to the terminal.

    Args:
        total: The total size of progress
        increase: Every time the default self-increase, the default is 1
        width: The width of the character output from the progress bar,
            the default is 50

    Expected display effect:
        [+][>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>]100.00%
        [-][>>>>>>>>>>>>>>>>>>>>>>>>>----------------]60.00%
    """
    def __init__(self, total, increase=1, width=50):
        self.total = total
        self.increase = increase
        self.width = width
        self.current = -1

    def output(self, current=None):
        """Output progress bar to terminal.

        Args:
            current: The current progress, if it is None, use the progress
            of self-increment.

        Note:
            When the first output is output, the default is 0.
        """
        if current is not None:
            self.current = current
        else:
            self.current += self.increase

        current_width = self.current * self.width // self.total
        remain_width = self.width - current_width
        current_percent = self.current * 100.0 / self.total

        out_format = '[{current}{remain}]{percent:.2f}%'
        if current_percent >= 100:
            current_width = self.width
            remain_width = 0
            current_percent = 100.00
            out_format = '\r[+]' + out_format
        else:
            out_format = '\r[-]' + out_format

        out = out_format.format(
            current = '>' * current_width,
            remain = '-' * remain_width,
            percent = current_percent
        )

        sys.stdout.write(out)
        sys.stdout.flush()

    def restart(self, total=None, inscrease=None, width=None):
        """Reinitialize the progress bar properties.

        Args:
            total: The total size of progress. If it is None,
                it will not be modified
            increase: Every time the default self-increase.
                If it is None, it will not be modified
            width: The width of the character output from the progress bar.
                If it is None, it will not be modified
        """
        if total is not None:
            self.total = total
        if inscrease is not None:
            self.increase = inscrease
        if width is not None:
            self.width = width
        self.current = -1

    def finish(self):
        """Complete output and return the instance object itself.
        """
        sys.stdout.write('\n')
        sys.stdout.flush()
        return self

File: c/progress_bar/progress_bar.c
/**
 * Copyright (c) 2018 by rgb-24bit.
 * License MIT, see LICENSE for more details.
 * ProgressBar use to output a simple progress bar to the terminal.
 */

#include <stdio.h>
#include <stdlib.h>

#include "progress_bar.h"

/* Create and initialize the ProgressBar structure. */
ProgressBar* create_progressbar(int total, int increase, int width) {
  ProgressBar* progressbar = (ProgressBar*)malloc(sizeof(ProgressBar));

  if (progressbar == NULL) {
    return NULL;
  }

  progressbar->total = total;
  progressbar->current = -1;
  progressbar->increase = increase == -1 ? 1 : increase;
  progressbar->width = width == -1 ? 50 : width;

  /* The longest output string: \r[+][width]100.00%\0 */
  progressbar->buf = (char*)malloc((progressbar->width + 20) * sizeof(char));

  if (progressbar->buf == NULL) {
    free(progressbar);
    return NULL;
  }

  return progressbar;
}

/* Destroy the ProgressBar structure. */
void destroy_progressbar(ProgressBar* progressbar) {
  if (progressbar != NULL) {
    if (progressbar->buf != NULL) {
      free(progressbar->buf);
    }
    free(progressbar);
  }
}

/* Build the string to be output. */
void __build_output_string(ProgressBar* progressbar) {
  float percent = progressbar->current * 100.0f / progressbar->total;

  char tail[9] = "]%.2f%%\0";

  if (percent >= 100) {
    sprintf(progressbar->buf, "\r[+][");

    for (int i = 0; i < progressbar->width; ++i) {
      progressbar->buf[5 + i] = '>';
    }

    percent = 100.0f;

  } else {
    sprintf(progressbar->buf, "\r[-][");

    int current_width = progressbar->width * percent / 100;
    for (int i = 0; i < current_width; ++i) {
      progressbar->buf[5 + i] = '>';
    }

    for (int i = current_width; i < progressbar->width; ++i) {
      progressbar->buf[5 + i] = '-';
    }
  }

  sprintf(progressbar->buf + 5 + progressbar->width, tail, percent);
}

/* Output progress bar to terminal. */
void output_progressbar(ProgressBar* progressbar, int current) {
  if (current == -1) {
    progressbar->current += progressbar->increase;
  } else {
    progressbar->current = current;
  }

  __build_output_string(progressbar);

  fputs(progressbar->buf, stdout);
}

/* finish output, output a line break to standard output */
ProgressBar* finish_output(ProgressBar* progressbar) {
  fputc('\n', stdout);
  return progressbar;
}

/* Reinitialize the progress bar properties */
ProgressBar* restart_progressbar(ProgressBar* progressbar,
                         int total, int increase, int width) {
  progressbar->current = -1;
  progressbar->total = total == -1 ? progressbar->total : total;
  progressbar->increase = increase == -1 ? progressbar->increase : increase;

  if (width != -1) {
    progressbar->buf = (char*)realloc(progressbar->buf,
                                      width + 20 * sizeof(char));

    if (progressbar->buf == NULL) {
      free(progressbar);
      return NULL;
    }

    progressbar->width = width;
  }

  return progressbar;
}
